### PR TITLE
Big bag of SSL fixes

### PR DESCRIPTION
**Pull request by alaviss**
This PR tackles the following issues with Nim's SSL implementation:

- `SSL_shutdown()` is called when it should not be (and causes `SSL_shutdown while in init` errors).
- The error queue is **not** cleared before an OpenSSL I/O operation. This can cause weird error to sprout out from nowhere.
- The `ALL` cipher list as used as the default. This cipher list may contain old and insecure ciphers.
- `getOpenSSLVersion()` doesn't support OpenSSL < 1.1.0
- `getOpenSSLVersion()` load symbols from the wrong library.

/cc @dom96 @FedericoCeratto 

Refs #13884

### CODE REVIEW

File: lib/wrappers/openssl.nim
Comment (alaviss) [inline lib/wrappers/openssl.nim]: @Araq @Clyybber I need help with this part, why is `{.gcsafe.}` triggered here?

Comment (alaviss) [reply lib/wrappers/openssl.nim]: In case you're wondering, removing `{.global.}` doesn't silence the warning.

**Comment by alaviss** [comment]
Removed `[backport]` due to the addition of `ssl_config` and the backward-incompatible change to `net` default cipher list.

As a side effect of the cipher list change, TLS < 1.2 support has now been practically disabled. Later PRs will introduce new APIs to support setting the lowest supported TLS version.

**Comment by alaviss** [comment]
I would like this PR to be merged as-is. The commit history contains a lot of important information that would aid people in hacking the SSL implementation

File: lib/pure/net.nim
Comment (alaviss) [inline lib/pure/net.nim]: This is added as a temporary fix for Windows. The OpenSSL DLLs for Windows should be updated to at least 1.0.2 (which is the oldest version that can be consider to be "secure" enough, though it's also reached EOL).

**Comment by alaviss** [comment]
If a rebase is wanted to clean up the history before merge, I can do that.

**Review by dom96 — APPROVED** [review]
This is awesome, thank you for working on it!

I really strongly feel that we need more test coverage here though. Right now there is very little and it's getting harder and harder to ensure this code works well, and in particular across different openssl versions. Would you be interested in improving this in follow up PRs?

Also, for the long-term it would be nice to make it so that any calls to OpenSSL are statically guaranteed to call ErrClearError. I bet you or someone in the future will miss a call to it. Could we maybe make use of the effect system here to ensure ErrClearError is called before each OpenSSL IO operation?

**Comment by FedericoCeratto** [comment]
There is also #13961 where we run security checks again a local server in Nim.

**Comment by alaviss** [comment]
> Also, for the long-term it would be nice to make it so that any calls to OpenSSL are statically guaranteed to call ErrClearError. I bet you or someone in the future will miss a call to it. Could we maybe make use of the effect system here to ensure ErrClearError is called before each OpenSSL IO operation?

I'm not quite sure how this can be done...

> I really strongly feel that we need more test coverage here though. Right now there is very little and it's getting harder and harder to ensure this code works well, and in particular across different openssl versions. Would you be interested in improving this in follow up PRs?

This is not something that I focus on, so I'll probably leave this to @FedericoCeratto. I've written a fair chunk of implementation information in the comments and the commit message, so everyone should be able to get a fair grip on how to work with all this code.

**Comment by timotheecour** [comment]
> I would like this PR to be merged as-is. The commit history contains a lot of important information that would aid people in hacking the SSL implementation

@alaviss the PR looks great; 
but next time please use `git rebase -i` to at least reduce down to just a few key commits. I doubt keeping all 28 commits actually helps for the next person hacking SSL, comments are better suited for that; also note that github still keeps all your commits if someone ever needs to look at your individual commits. Imagine if all PR's did that; it adds friction for git rebase (more chances of conflict), for git bisect (intermediate broken state depending on your git bisect test) and for following git log `<end of nag>` /cc @Araq

**Comment by alaviss** [comment]
> but next time please use `git rebase -i` to at least reduce down to just a few key commits.

Fake history is not history, and not all commits can be reduced this way. You can try to reduce commits to "look nice and smart", but can you spend the time to make sure that each commits you reduced work and have up-to-date description? It's not uncommon for commits to describe something unrelated to the actual change due to this practice.

> I doubt keeping all 28 commits actually helps for the next person hacking SSL, comments are better suited for that

Comments describe the code, commits is how you get there. I can document every `ErrClearError()` call I added there with the same message, or I can use a commit as my documentation which neatly describes why it's done in all those exact locations.

When you need to do a `git blame` trace to see why a decision was made, you look at the commits and the code it touches, and this become a pain when something is compressed into a few commits for "clean history". You can't isolate why a decision was made and you have to guess. In fact, some of the commit-then-revert you see here was a direct result of it happening to the history of a simple `net.nim` change. I have to guess what the older commits were trying to do because they were compressed into one bundle without comments of any form.

Please do note that I don't request things like this unless I believe that the history is of adequate quality and that it's worth keeping it.

> also note that github still keeps all your commits if someone ever needs to look at your individual commits.

Sure it might keep it, but will anyone actually look when they do `git blame` on their local device? Or will they grumble about how this one big change touches every single file and viewing the diff suddenly becomes a chore?

Even a small change to one line become a pain to track down when it's a part of a commit touching multiple files, because then you have absolutely no idea if the change was crucial or was just a "small tweak along the way" because the developer didn't like how the code look at the moment. I've encountered this more than I'd like.

**Comment by timotheecour** [comment]
> Fake history is not history

I've heard similar arguments to justify git merge to instead of git rebase to preserve history of when a patch was merged; it turns out that this aspect of history is only relevant for the person writing the PR and is irrelevant once PR is merged in main branch.

as for keeping intermediate commits: "Happy families are all alike; every unhappy family is unhappy in its own way". Probably a bit far fetched analogy but what I mean by that is there are many ways to get to a happy state, and in the end it doesn't really matter which path was taken and which regressions/glitches/bugs were introduced encountered along the way so long the end result is clear code, with comments inserted where needed if something subtle needs to be mentioned (and PR comments + commits are there too)

intermediate commits are fine if they fix separate topics but ideally should be redacted to go from "sane state" to "sane state"

> It's not uncommon for commits to describe something unrelated to the actual change due to this practice

by default, `git rebase -i` + will aggregate your individual commits marked as `f` (but oftentimes still a good idea to do cleanups)

also git blame doesn't survive well a few refactorings or nimpretty invocations, unlike comments in code

> one big change touches every single file and viewing the diff suddenly becomes a chore?

sum(diff) > diff(sum) because you're touching some lines multiple times; this makes rebase harder; for the same reason, oftentimes it's easier to squash (or reduce) before rebase

anyway, it's fine and benefits of this PR outweigh this so I'll stop here

**Comment by Araq** [comment]
> Even a small change to one line become a pain to track down when it's a part of a commit touching multiple files, because then you have absolutely no idea if the change was crucial or was just a "small tweak along the way" because the developer didn't like how the code look at the moment. I've encountered this more than I'd like.

It's probably my fault and I'm sorry. And I'm not doing that as often anymore as I used to. However, source code is better when you don't have to read git's meta information at all and it's written down in comments what didn't work and why. Just yesterday I had an annoying ``git log -p`` session in attempt to remember a subtle detail in ``sigmatch.nim``. More dead code with ``when false: # comment`` would have sped up that process.

**Pull request merged**
